### PR TITLE
Vararg fixes byval

### PR DESCRIPTION
**Pull request by pointhi**
Combines #717 and #719, which is required so we were able to implement the required code to pass structs via varargs.

**Changes:**
* add parameter support to call and invoke instructions
* implement parameter attribute parsing for llvm 3.2 bitcode (required for gcc-tests)
* Add interface to allow general access for those information
@grimmerm: please check if this interface declaration doesn't interfere with design decisions regarding call and invoke.
* cleanup and fix vararg handling to be x86_64 compatible
  * Fix: #711
  * Fix: #716
* add additional testcases, and enable some old ones where the behaviour is now fixed

### CODE REVIEW

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java [hunk 1]
```diff
@@ -108,73 +126,124 @@ private FrameSlot getStackPointerSlot() {
...
+        unboxArguments(realArguments, numberOfExplicitArguments);
+        final int nrVarArgs = realArguments.length - numberOfExplicitArguments;
 
-        // Init reg_save_area:
-        // #############################
-        // Allocate worst amount of memory - saves a few ifs
+        // Allocate register_save_area
         LLVMAddress structAddress = targetToAddress.executeWithTarget(target.executeGeneric(frame));
-        long regSaveArea = LLVMStack.allocateStackMemory(frame, getStackPointerSlot(), X86_64BitVarArgs.GP_LIMIT + X86_64BitVarArgs.FP_LIMIT, 8);
+        final long regSaveArea = LLVMStack.allocateStackMemory(frame, getStackPointerSlot(), X86_64BitVarArgs.FP_LIMIT, 8);
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java]: we normally do not declare function local variables as final

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java [hunk 2]
```diff
@@ -108,73 +126,124 @@ private FrameSlot getStackPointerSlot() {
...
+            VarArgArea area = getVarArgArea(realArguments[i]);
+            if (area == VarArgArea.GP_AREA && gpOffset < X86_64BitVarArgs.GP_LIMIT) {
+                gpOffset += X86_64BitVarArgs.GP_STEP;
+            } else if (area == VarArgArea.FP_AREA && fpOffset < X86_64BitVarArgs.FP_LIMIT) {
+                fpOffset += X86_64BitVarArgs.FP_STEP;
+            } else if (area != VarArgArea.OVERFLOW_AREA) {
+                overflowArea += X86_64BitVarArgs.STACK_STEP;
+            } else {
+                // TODO: we need to know the required stacksize
+                overflowArea += 128;
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java]: what's that?

Comment (pointhi) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java]: oh, that's a workaround before the time I knew the stacksize. I will fix it immediatly

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java [hunk 3]
```diff
@@ -471,6 +472,12 @@ public Object doAddress(LLVMAddress address, LLVMAddress value) {
         }
 
         @Specialization
+        public Object doAddress(LLVMAddress address, LLVMVarArgCompoundValue value) {
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java]: how can that value end up in this node?

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java [hunk 4]
```diff
@@ -471,6 +472,12 @@ public Object doAddress(LLVMAddress address, LLVMAddress value) {
...
+            LLVMMemory.putAddress(address, value.getAddr());
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java]: isn't the semantic of a compound argument value the be always copied by value?

File: projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/attributes/AtributeGroupGetter.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.parser.model.attributes;
+
+public interface AtributeGroupGetter {
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/attributes/AtributeGroupGetter.java]: Typo

Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/attributes/AtributeGroupGetter.java]: And a bad name for an interface, a "getter" is a function

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java (hunk 1 above)
Comment (pointhi) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/x86/LLVMX86_64BitVAStart.java]: If you want I can change it to the non final version.

File: projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java (hunk 3 above)
Comment (pointhi) [inline projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/memory/LLVMStoreNode.java]: seems to be a leftover of some older implementation, I will remove it

File: projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/FunctionDeclaration.java [hunk 6]
```diff
@@ -38,7 +39,7 @@
 import com.oracle.truffle.llvm.runtime.types.symbols.LLVMIdentifier;
 import com.oracle.truffle.llvm.runtime.types.symbols.ValueSymbol;
 
-public final class FunctionDeclaration implements Constant, ValueSymbol {
+public final class FunctionDeclaration implements Constant, ValueSymbol, AtributeGroupGetter {
```
Comment (grimmerm) [inline projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/FunctionDeclaration.java]: Is it really necessary to introduce another interface?

Comment (pointhi) [inline projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/FunctionDeclaration.java]: Hm, I think it should also work to move the methods of ```AtributeGroupGetter``` into the ```Call``` and ```Invoke``` Interface. Is probably the cleaner solution, even we have a little bit of duplication then.

**Pull request closed without merging**